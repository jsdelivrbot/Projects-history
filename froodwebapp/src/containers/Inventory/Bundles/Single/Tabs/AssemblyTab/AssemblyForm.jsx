/* eslint-disable jsx-a11y/label-has-for */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Row, Col } from 'react-flexbox-grid';
import { connect } from 'react-redux';
import {
  FormInputNumber,
  Button,
} from 'components';
import {
  bundleAssembliesGetRequest,
  bundleAssembliesUpdateRequest,
} from 'redux-base/actions';
import { reduxForm, Field } from 'redux-form';
import { form } from 'styles/common.scss';
import { assembledButton, disassembledButton } from './AssemblyForm.scss';

const reduxFormConfig = {
  form: 'itemBundleAssemblyForm',
  enableReinitialize: true
};

const mapStateToProps = state => ({
  initialValues: state.bundle.bundleAssemblies
});

const mapDispatchToProps = {
  bundleAssembliesGetRequest,
  bundleAssembliesUpdateRequest,
};

export class AssemblyForm extends Component {
  handleUpdate = (values) => {
    const doAssemble = this.props.initialValues.assembledQty === 0;
    const payload = {
      qty: doAssemble ? values.assembledQty : values.disassembledQty,
      operation: doAssemble ? 'assemble' : 'disassemble'
    };

    if (doAssemble) {
      payload.id = this.props.initialValues.id;
    }

    this.props.bundleAssembliesUpdateRequest({
      id: this.props.bundleId,
      payload
    });
  }

  render() {
    const {
      initialValues,
      handleSubmit
    } = this.props;

    const showAssemble = initialValues.assembledQty === 0;

    return (
      <div>
        <form
          onSubmit={ handleSubmit(this.handleUpdate) }
          className={ form }
        >
          <Row bottom="xs">
            <Col xs={ 12 } lg={ 5 }>
              <Row middle="xs" center="xs">
                <Col xs={ 5 }>
                  <label>Max Bundle Quantity</label>
                </Col>
                <Col xs={ 7 }>
                  <Field
                    name="maxBundleQty"
                    min={ 0 }
                    disabled
                    component={ FormInputNumber }
                  />
                </Col>
              </Row>
              <Row middle="xs" center="xs">
                <Col xs={ 5 }>
                  <label>Assembled Bundle Quantity</label>
                </Col>
                <Col xs={ 7 }>
                  <Field
                    name="assembledQty"
                    disabled={ !showAssemble }
                    max={ initialValues.maxBundleQty }
                    min={ 0 }
                    component={ FormInputNumber }
                  />
                </Col>
              </Row>
              { !showAssemble &&
                <Row middle="xs" center="xs">
                  <Col xs={ 5 }>
                    <label>Disassembled Bundle Quantity</label>
                  </Col>
                  <Col xs={ 7 }>
                    <Field
                      name="disassembledQty"
                      max={ initialValues.assembledQty }
                      min={ 0 }
                      component={ FormInputNumber }
                    />
                  </Col>
                </Row>
              }
            </Col>
            <Col lg={ 2 }>
              <Row>
                <Col>
                  <Button
                    type="submit"
                    className={ showAssemble ? assembledButton : disassembledButton }
                  >
                    { showAssemble ? 'Assemble' : 'Disassemble' }
                  </Button>
                </Col>
              </Row>
            </Col>
          </Row>
        </form>
      </div>
    );
  }
}

AssemblyForm.propTypes = {
  // redux-form related props
  handleSubmit: PropTypes.func.isRequired,
  // props
  bundleId: PropTypes.string.isRequired,
  initialValues: PropTypes.object.isRequired,
  // redux-base
  bundleAssembliesUpdateRequest: PropTypes.func.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(reduxForm(reduxFormConfig)(AssemblyForm));
