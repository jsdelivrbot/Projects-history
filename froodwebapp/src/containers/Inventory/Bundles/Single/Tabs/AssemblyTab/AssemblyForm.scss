@import '~styles/colors.scss';

.disassembledButton {
  border: $border-width solid $cancel-button-color;
  color: $cancel-button-color;
}

.assembledButton {
  border: $border-width solid $save-button-color;
  color: $save-button-color;
}
