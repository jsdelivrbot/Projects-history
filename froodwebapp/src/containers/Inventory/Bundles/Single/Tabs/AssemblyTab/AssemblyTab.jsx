import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import {
  bundleAssembliesGetRequest
} from 'redux-base/actions';
import { Table, Spin } from 'antd';
import { table } from 'styles/common.scss';
import ConnectedAssemblyForm from './AssemblyForm';
import columns from './assemblyTabHelpers';

const mapStateToProps = state => ({
  loadingPage: state.bundle.loadingPage,
  needReloadAssemblies: state.bundle.needReloadAssemblies,
  bundleAssembliesDetails: state.bundle.bundleAssemblies.details,
});

const mapDispatchToProps = {
  bundleAssembliesGetRequest
};

export class AssemblyTab extends Component {

  componentWillReceiveProps(nextProps) {
    // when we update bundle assemblies
    if (nextProps.needReloadAssemblies) {
      this.props.bundleAssembliesGetRequest({
        id: this.props.bundleId
      });
    }
  }

  render() {
    const {
      bundleAssembliesDetails,
      loadingPage,
      bundleId
    } = this.props;

    return (
      <div>
        <Spin spinning={ loadingPage }>
          <ConnectedAssemblyForm
            bundleId={ bundleId }
          />
          <Table
            className={ table }
            rowKey="id"
            size="small"
            columns={ columns }
            dataSource={ bundleAssembliesDetails }
          />
        </Spin>
      </div>
    );
  }
}

AssemblyTab.propTypes = {
  // trigger
  loadingPage: PropTypes.bool.isRequired,
  needReloadAssemblies: PropTypes.bool.isRequired,
  // data
  bundleId: PropTypes.string.isRequired,
  bundleAssembliesDetails: PropTypes.array.isRequired,
  // redux-base
  bundleAssembliesGetRequest: PropTypes.func.isRequired,
};

export default connect(mapStateToProps, mapDispatchToProps)(AssemblyTab);
