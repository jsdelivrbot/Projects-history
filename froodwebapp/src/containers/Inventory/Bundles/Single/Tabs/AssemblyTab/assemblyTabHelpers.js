export default [{
  title: 'SKU',
  dataIndex: 'sku',
}, {
  title: 'Name',
  dataIndex: 'skuName',
}, {
  title: 'Category',
  dataIndex: 'categoryName',
}, {
  title: 'UOM',
  dataIndex: 'uomName',
}, {
  title: 'Batch',
  dataIndex: 'batch',
}, {
  title: 'Expiry',
  dataIndex: 'expiry',
}, {
  title: 'Quantity',
  dataIndex: 'qty',
}];
