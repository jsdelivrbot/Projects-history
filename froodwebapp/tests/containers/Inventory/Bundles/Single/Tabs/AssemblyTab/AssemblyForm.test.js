import { AssemblyForm } from 'containers/Inventory/Bundles/Single/Tabs/AssemblyTab/AssemblyForm';
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

const { describe, expect, it } = global;

describe('AssemblyForm', () => {
  const props = {
    loadingPage: false,
    bundleId: '1',
    handleSubmit: jest.fn(),
    bundleAssembliesUpdateRequest: jest.fn(),
    initialValues: {
      id: 1,
      assembledQty: 10,
      maxBundleQty: 25,
    }
  };
  it('renders the AssemblyForm component', () => {
    const component = shallow(<AssemblyForm { ...props } />);

    expect(component.length).toEqual(1);
  });

  it('capturing Snapshot of AssemblyForm', () => {
    const component = shallow(<AssemblyForm { ...props } />);
    const view = toJson(component);
    const actual = expect(view);

    actual.toMatchSnapshot();
  });

  it('calls bundleAssembliesUpdateRequest in handleUpdate with correct param', () => {
    const values = {
      assembledQty: 14,
      disassembledQty: 24
    }
    const component = shallow(<AssemblyForm { ...props } />);
    component.instance().handleUpdate(values);

    expect(props.bundleAssembliesUpdateRequest.mock.calls.length).toBe(1);

    expect(props.bundleAssembliesUpdateRequest.mock.calls[0][0]).toEqual({
      id: props.bundleId,
      payload: {
        qty: 24,
        operation: 'disassemble'
      }
    });
  });
});
