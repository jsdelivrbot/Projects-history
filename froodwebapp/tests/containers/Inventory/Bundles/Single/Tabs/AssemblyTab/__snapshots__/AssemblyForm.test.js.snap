// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssemblyForm capturing Snapshot of AssemblyForm 1`] = `
<div>
  <form
    className="form"
  >
    <Row
      bottom="xs"
    >
      <Col
        lg={5}
        xs={12}
      >
        <Row
          center="xs"
          middle="xs"
        >
          <Col
            xs={5}
          >
            <label>
              Max Bundle Quantity
            </label>
          </Col>
          <Col
            xs={7}
          >
            <Field
              component={[Function]}
              disabled={true}
              min={0}
              name="maxBundleQty"
            />
          </Col>
        </Row>
        <Row
          center="xs"
          middle="xs"
        >
          <Col
            xs={5}
          >
            <label>
              Assembled Bundle Quantity
            </label>
          </Col>
          <Col
            xs={7}
          >
            <Field
              component={[Function]}
              disabled={true}
              max={25}
              min={0}
              name="assembledQty"
            />
          </Col>
        </Row>
        <Row
          center="xs"
          middle="xs"
        >
          <Col
            xs={5}
          >
            <label>
              Disassembled Bundle Quantity
            </label>
          </Col>
          <Col
            xs={7}
          >
            <Field
              component={[Function]}
              max={10}
              min={0}
              name="disassembledQty"
            />
          </Col>
        </Row>
      </Col>
      <Col
        lg={2}
      >
        <Row>
          <Col>
            <Button
              className="disassembledButton"
              type="submit"
            >
              Disassemble
            </Button>
          </Col>
        </Row>
      </Col>
    </Row>
  </form>
</div>
`;
